### PR TITLE
group snippets with the same data-call-uri to reduce number of calls …

### PR DESCRIPTION
**Pull request by michalwronowski**
…to microservices

### CODE REVIEW

File: src/main/java/com/cognifide/knotx/handler/RestServiceResponseHandler.java [hunk 1]
```diff
@@ -70,19 +66,25 @@ public void handle(HttpClientResponse response) {
         response.bodyHandler(buffer -> {
             String responseContent = buffer.getString(0, buffer.length());
             LOGGER.debug("Request in: " + request.absoluteURI() + " for " + dataCallUri);
+            Map serviceData = new Gson().fromJson(responseContent, Map.class);
```
Comment (toniedzwiedz) [inline src/main/java/com/cognifide/knotx/handler/RestServiceResponseHandler.java]: You can avoid using the raw `Map` type by specifying a type token as explained in the [GSON FAQ](https://sites.google.com/site/gson/gson-user-guide#TOC-Serializing-and-Deserializing-Generic-Types)

File: src/main/java/com/cognifide/knotx/template/TemplateHandlerImpl.java [hunk 2]
```diff
@@ -97,21 +105,28 @@ public void finishIfLast(HttpServerRequest request) {
...
+
+    private com.github.jknack.handlebars.Template compile(String html, String dataCallUri) {
         try {
-            com.github.jknack.handlebars.Template template = handlebars.compileInline(templateContent);
-            RestServiceResponseHandler serviceResponseHandler =
-                    new RestServiceResponseHandler(request, template, this, dataCallUri, observableRequest, snippet, templateDebug);
-            server.callService(request, dataCallUri, serviceResponseHandler);
+            return handlebars.compileInline(html);
         } catch (IOException e) {
             LOGGER.error("Could not process template [{}]", dataCallUri);
```
Comment (toniedzwiedz) [inline src/main/java/com/cognifide/knotx/template/TemplateHandlerImpl.java]: Wouldn't it be useful to also log the root cause? `e` is never used here.

File: src/main/java/com/cognifide/knotx/template/TemplateHandlerImpl.java [hunk 3]
```diff
@@ -97,21 +105,28 @@ public void finishIfLast(HttpServerRequest request) {
...
-    private void handleTemplate(Element snippet, HttpServerRequest request) {
-        final String dataCallUri = getServiceUrl(request, snippet);
+    private void handleTemplate(Entry<String, List<Element>> snippetGroup, HttpServerRequest request) {
+        final String dataCallUri = snippetGroup.getKey();
         ObservableRequest observableRequest = new ObservableRequest(dataCallUri);
         observableRequest.addObserver(trafficObserver);
         observableRequest.onStart();
 
-        String templateContent = snippet.html();
+        Map<Element, com.github.jknack.handlebars.Template> snippetTemplateMap = snippetGroup.getValue().stream()
```
Comment (Skejven) [inline src/main/java/com/cognifide/knotx/template/TemplateHandlerImpl.java]: Not 100% sure if 'org.jsoup.nodes.Element' is good candidate for map key?

**Comment by tomaszmichalak** [comment]
Please update documentation - it is very important feature. It will be good to draw diagram with explanation.

File: README.md [hunk 4]
```diff
@@ -187,6 +187,43 @@ Another way to provide a value for the password placeholder shown above is to se
 
 >Notice: command line arguments take precedence over environment variables.
 
+# Features
+
+## Requests grouping
+
+Template obtained from the repository may containt many snippets that will trigger microservice calls for data. There is a chance that some of the snippets will have the same `data-call-uri` attribute set, meaning they will request data from the same source.
```
Comment (toniedzwiedz) [inline README.md]: Typo in `contain`

File: README.md [hunk 5]
```diff
@@ -187,6 +187,43 @@ Another way to provide a value for the password placeholder shown above is to se
...
+Let's assume that we obtained following template from repository:
+```html
+<div>
+<script data-api-type="templating" data-call-uri="/searchService" type="text/x-handlebars-template">
+    <div>{{search.term}}</div>
+</script>
+</div>
+...
+<div>
+<script data-api-type="templating" data-call-uri="/searchService" type="text/x-handlebars-template">
```
Comment (toniedzwiedz) [inline README.md]: `the following`

File: README.md [hunk 6]
```diff
@@ -187,6 +187,43 @@ Another way to provide a value for the password placeholder shown above is to se
...
 
+# Features
+
+## Requests grouping
+
+Template obtained from the repository may containt many snippets that will trigger microservice calls for data. There is a chance that some of the snippets will have the same `data-call-uri` attribute set, meaning they will request data from the same source.
+In such case only one call to miscroservice shall be made and data retrieved from service call should be applied to all snippets sharing the same `data-call-uri`.
+
+Example:
+Let's assume that we obtained following template from repository:
```
Comment (toniedzwiedz) [inline README.md]: `the following`

**Comment by tomaszmichalak** [comment]
All changes approved.

**Pull request merged**